core/src/main/java/fj/data/NonEmptyList.java

```diff
@@ -395,13 +378,540 @@ public final A maximum(final Ord<A> o) {
    * @param obj the other object to check for equality against.
    * @return true if this list is equal to the provided argument
    */
-  @Override public boolean equals( final Object obj ) {
+  @Override
+  public boolean equals(final Object obj) {
     return Equal.equals0(NonEmptyList.class, this, obj, () -> Equal.nonEmptyListEqual(Equal.anyEqual()));
   }
 
-  @Override public int hashCode() {
+  @Override
+  public int hashCode() {
     return Hash.nonEmptyListHash(Hash.<A>anyHash()).hash(this);
   }
 
-  @Override public String toString() { return Show.nonEmptyListShow(Show.<A>anyShow()).showS(this); }
+  @Override
+  public String toString() {
+    return Show.nonEmptyListShow(Show.<A>anyShow()).showS(this);
+  }
+
+  /**
+   * Fold the list, from right to left.
+   *
+   * @param combine the combine function
+   * @return the output
+   */
+  public final A foldRight1(
+          final F<A, F<A, A>> combine) {
+    return foldRightN(combine, identity());
+  }
+
+  /**
+   * Fold the list, from right to left.
+   *
+   * @param combine the combine function
+   * @return the output
+   */
+  public final A foldRight1(
+          final F2<A, A, A> combine) {
+    return foldRightN(combine, identity());
+  }
+
+  /**
+   * Fold the list, from left to right.
+   *
+   * @param combine the combine function
+   * @return the output
+   */
+  public final A foldLeft1(
+          final F<A, F<A, A>> combine) {
+    return foldLeftN(combine, identity());
+  }
+
+  /**
+   * Fold the list, from left to right.
+   *
+   * @param combine   the combine function
+   * @return the output
+   */
+  public final A foldLeft1(
+          final F2<A, A, A> combine) {
+    return foldLeftN(combine, identity());
+  }
+
+  /**
+   * Fold the list, from right to left, using one function to transform the last and another function to combine the
+   * output.
+   *
+   * @param combine   the combine function
+   * @param transform the transform function
+   * @param <B>       the type of the output
+   * @return the output
+   */
+  public final <B> B foldRightN(
+          final F<A, F<B, B>> combine,
+          final F<A, B> transform) {
+    return tail().isEmpty() ?
+            transform.f(head()) :
+            init().foldRight(combine, transform.f(last()));
```

Comment:
I think this is overly inefficient.  init is O(n), last is O(n) and foldRight is O(n), so we have overall O(n^2).  Compare this to List foldRight, which is O(n) for the reverse and then O(n) for foldLeft, for an overall O(n).

Reply:
I think we should be reversing and then folding left,